### DISCUSSION TITLE
Read User Profile API have some data discrepancies

### DISCUSSION THREAD

**Discussion by Shakthieshwari** [Sunbird-Lern/Community#6] [Issues]
Hello,

On Sunbird , we have read user api :-
End Point - https://staging.sunbirded.org/learner/user/v5/read/{{user_id}}

We could find issue with the response we get from this api :-
Response :-
1. {"id":"api.user.read.7f93a6be-1eb7-4cbe-a616-40d4c4ae8be3","ver":"v5","ts":"2022-05-26 09:49:26:985+0000","params":{"resmsgid":"8eaa102b-180a-0b81-5409-38655643b380","msgid":"8eaa102b-180a-0b81-5409-38655643b380","err":null,"status":"SUCCESS","errmsg":null},"responseCode":"OK","result":{"response":{"maskedPhone":null,"channel":"dikshapreprodcustodian","profileUserTypes":[{"type":"administrator","subType":"dct"},{"type":"administrator","subType":"hm"}],"updatedDate":"2022-05-26 09:49:25:322+0000","managedBy":null,"flagsValue":0,"id":"7f93a6be-1eb7-4cbe-a616-40d4c4ae8be3","recoveryEmail":"","identifier":"7f93a6be-1eb7-4cbe-a616-40d4c4ae8be3","updatedBy":"7f93a6be-1eb7-4cbe-a616-40d4c4ae8be3","externalIds":[],"roleList":[{"name":"Book Creator","id":"BOOK_CREATOR"},{"name":"Membership Management","id":"MEMBERSHIP_MANAGEMENT"},{"name":"Flag Reviewer","id":"FLAG_REVIEWER"},{"name":"Report Viewer","id":"REPORT_VIEWER"},{"name":"Program Manager","id":"PROGRAM_MANAGER"},{"name":"Program Designer","id":"PROGRAM_DESIGNER"},{"name":"System Administration","id":"SYSTEM_ADMINISTRATION"},{"name":"Content Curation","id":"CONTENT_CURATION"},{"name":"Book Reviewer","id":"BOOK_REVIEWER"},{"name":"Content Creator","id":"CONTENT_CREATOR"},{"name":"Org Management","id":"ORG_MANAGEMENT"},{"name":"Public","id":"PUBLIC"},{"name":"Course Mentor","id":"COURSE_MENTOR"},{"name":"Content Reviewer","id":"CONTENT_REVIEWER"},{"name":"Report Admin","id":"REPORT_ADMIN"},{"name":"Org Admin","id":"ORG_ADMIN"}],"rootOrgId":"0126796199493140480","prevUsedEmail":"","firstName":"rio0033","isMinor":false,"tncAcceptedOn":1653303725644,"allTncAccepted":{},"profileDetails":null,"phone":"","dob":"2000-12-31","status":1,"lastName":"","tncLatestVersion":"v12","roles":[],"prevUsedPhone":"","stateValidated":false,"isDeleted":false,"organisations":[{"organisationId":"0126796199493140480","approvedBy":null,"channel":"dikshapreprodcustodian","updatedDate":null,"approvaldate":null,"isSystemUpload":false,"isDeleted":false,"id":"01354387056744038431","isApproved":null,"orgjoindate":"2022-05-23 11:02:01:905+0000","isSelfDeclaration":true,"updatedBy":null,"orgName":"Pre-prod Custodian Organization","addedByName":null,"addedBy":null,"associationType":2,"locationIds":["027f81d8-0a2c-4fc6-96ac-59fe4cea3abf","8250d58d-f1a2-4397-bfd3-b2e688ba7141"],"orgLocation":[{"type":"state","id":"027f81d8-0a2c-4fc6-96ac-59fe4cea3abf"},{"type":"district","id":"8250d58d-f1a2-4397-bfd3-b2e688ba7141"}],"externalId":null,"userId":"7f93a6be-1eb7-4cbe-a616-40d4c4ae8be3","isSchool":false,"hashTagId":"0126796199493140480","isSSO":false,"isRejected":null,"locations":[{"code":"29","name":"Karnataka","id":"027f81d8-0a2c-4fc6-96ac-59fe4cea3abf","type":"state","parentId":null},{"code":"2901","name":"BELAGAVI","id":"8250d58d-f1a2-4397-bfd3-b2e688ba7141","type":"district","parentId":"027f81d8-0a2c-4fc6-96ac-59fe4cea3abf"}],"position":null,"orgLeftDate":null},{"organisationId":"013348013296558080195","approvedBy":null,"channel":"up","updatedDate":"2022-05-26 09:49:25:333+0000","approvaldate":null,"isSystemUpload":false,"isDeleted":false,"id":"013348013296558080195","isApproved":false,"orgjoindate":"2022-05-25 12:26:22:087+0000","isSelfDeclaration":true,"updatedBy":"7f93a6be-1eb7-4cbe-a616-40d4c4ae8be3","orgName":"JHS NAGLA SADLE COMPOSITE","addedByName":null,"addedBy":"7f93a6be-1eb7-4cbe-a616-40d4c4ae8be3","associationType":2,"locationIds":["a87aa157-0d73-4aa3-9623-cb2f8a3ecf32","6d884bb0-307f-4f83-abfe-fc21bbd36abb","646a627f-f024-494c-828e-f229c2e9f566"],"orgLocation":[{"id":"a87aa157-0d73-4aa3-9623-cb2f8a3ecf32","type":"district"},{"id":"6d884bb0-307f-4f83-abfe-fc21bbd36abb","type":"state"},{"id":"646a627f-f024-494c-828e-f229c2e9f566","type":"block"}],"externalId":"09150206302","userId":"7f93a6be-1eb7-4cbe-a616-40d4c4ae8be3","isSchool":true,"hashTagId":"013348013296558080195","isSSO":false,"isRejected":false,"locations":[{"code":"09146","name":"AGRA","id":"a87aa157-0d73-4aa3-9623-cb2f8a3ecf32","type":"district","parentId":"6d884bb0-307f-4f83-abfe-fc21bbd36abb"},{"code":"09","name":"Uttar Pradesh","id":"6d884bb0-307f-4f83-abfe-fc21bbd36abb","type":"state","parentId":null},{"code":"09001754","name":"ACHHNERA","id":"646a627f-f024-494c-828e-f229c2e9f566","type":"block","parentId":"a87aa157-0d73-4aa3-9623-cb2f8a3ecf32"}],"position":null,"orgLeftDate":null}],"countryCode":null,"tncLatestVersionUrl":"https://preprodall.blob.core.windows.net/termsandcond/terms-and-conditions-v12.html","maskedEmail":"ri*****@yopmail.com","email":"ri*****@yopmail.com","rootOrg":{"keys":{"encKeys":["456"],"signKeys":["456"]},"organisationSubType":null,"channel":"dikshapreprodcustodian","description":"Pre-prod Custodian Organization","updatedDate":"2020-08-28 10:12:01:096+0000","organisationType":5,"isTenant":true,"provider":null,"id":"0126796199493140480","isBoard":true,"email":null,"slug":"dikshapreprodcustodian","isSSOEnabled":null,"orgName":"Pre-prod Custodian Organization","updatedBy":null,"locationIds":["027f81d8-0a2c-4fc6-96ac-59fe4cea3abf","8250d58d-f1a2-4397-bfd3-b2e688ba7141"],"externalId":null,"orgLocation":[{"type":"state","id":"027f81d8-0a2c-4fc6-96ac-59fe4cea3abf"},{"type":"district","id":"8250d58d-f1a2-4397-bfd3-b2e688ba7141"}],"isRootOrg":true,"rootOrgId":"0126796199493140480","imgUrl":null,"homeUrl":null,"createdDate":"2019-01-18 09:48:13:428+0000","createdBy":"system","hashTagId":"0126796199493140480","status":1},"userLocations":[{"code":"09001754","name":"ACHHNERA","id":"646a627f-f024-494c-828e-f229c2e9f566","type":"block","parentId":"a87aa157-0d73-4aa3-9623-cb2f8a3ecf32"},{"code":"09","name":"Uttar Pradesh","id":"6d884bb0-307f-4f83-abfe-fc21bbd36abb","type":"state","parentId":null},{"code":"09146","name":"AGRA","id":"a87aa157-0d73-4aa3-9623-cb2f8a3ecf32","type":"district","parentId":"6d884bb0-307f-4f83-abfe-fc21bbd36abb"},{"code":"09150206302","name":"JHS NAGLA SADLE COMPOSITE","id":"013348013296558080195","type":"school","parentId":""}],"recoveryPhone":"","userName":"rio0033_syrk","userId":"7f93a6be-1eb7-4cbe-a616-40d4c4ae8be3","declarations":[],"promptTnC":false,"lastLoginTime":0,"createdDate":"2022-05-23 11:02:01:884+0000","framework":{"board":["State (Uttar Pradesh)"],"gradeLevel":["Class 3"],"id":["up_k-12"],"medium":["ENGLISH"],"subject":["English"]},"createdBy":null,"profileUserType":{"subType":"dct","type":"administrator"},"tncAcceptedVersion":"v12"}}}

**Inside the organisations array in the above response, we could find both school and organisations details why ? we can differentiate an org with "isSchool":false. basically organisations array should contain only organisation data, why is the school data coming into it?** 
Cc - @aks30 @kiranharidas187 @vijiurs

2. {"id":"api.user.read.3f8861d6-37a6-493a-8cdc-ca23b27f472c","ver":"v5","ts":"2022-05-26 07:15:25:336+0000","params":{"resmsgid":"12505a3d-7e9e-18c8-0fa4-fa58924962ac","msgid":"12505a3d-7e9e-18c8-0fa4-fa58924962ac","err":null,"status":"SUCCESS","errmsg":null},"responseCode":"OK","result":{"response":{"maskedPhone":null,"tcStatus":null,"channel":"tn","profileUserTypes":[{"type":"administrator","subType":"hm"},{"type":"administrator","subType":"crp"}],"updatedDate":"2022-02-22 07:41:54:216+0000","managedBy":null,"flagsValue":4,"id":"3f8861d6-37a6-493a-8cdc-ca23b27f472c","recoveryEmail":"","identifier":"3f8861d6-37a6-493a-8cdc-ca23b27f472c","updatedBy":"3f8861d6-37a6-493a-8cdc-ca23b27f472c","externalIds":[],"roleList":[{"name":"Book Creator","id":"BOOK_CREATOR"},{"name":"Membership Management","id":"MEMBERSHIP_MANAGEMENT"},{"name":"Flag Reviewer","id":"FLAG_REVIEWER"},{"name":"Report Viewer","id":"REPORT_VIEWER"},{"name":"Program Manager","id":"PROGRAM_MANAGER"},{"name":"Program Designer","id":"PROGRAM_DESIGNER"},{"name":"System Administration","id":"SYSTEM_ADMINISTRATION"},{"name":"Content Curation","id":"CONTENT_CURATION"},{"name":"Book Reviewer","id":"BOOK_REVIEWER"},{"name":"Content Creator","id":"CONTENT_CREATOR"},{"name":"Org Management","id":"ORG_MANAGEMENT"},{"name":"Course Admin","id":"COURSE_ADMIN"},{"name":"Org Moderator","id":"ORG_MODERATOR"},{"name":"Public","id":"PUBLIC"},{"name":"Admin","id":"ADMIN"},{"name":"Course Mentor","id":"COURSE_MENTOR"},{"name":"Content Reviewer","id":"CONTENT_REVIEWER"},{"name":"Report Admin","id":"REPORT_ADMIN"},{"name":"Org Admin","id":"ORG_ADMIN"}],"rootOrgId":"01269878797503692810","prevUsedEmail":"","firstName":"radmin","tncAcceptedOn":1645005384815,"allTncAccepted":{"reportViewerTnc":{"tncAcceptedOn":"2022-02-23 11:03:03:002+0000","version":"4.5.0"}},"profileDetails":null,"phone":"","dob":null,"status":1,"lastName":"","tncLatestVersion":"v12","roles":[{"role":"REPORT_ADMIN","createdDate":"2022-02-16 09:59:38:904+0000","updatedBy":null,"createdBy":"fbe926ac-a395-40e4-a65b-9b4f711d7642","scope":[{"organisationId":"01269878797503692810"}],"updatedDate":null}],"prevUsedPhone":"","stateValidated":true,"isDeleted":false,"organisations":[{"organisationId":"01269878797503692810","approvedBy":null,"channel":"tn","updatedDate":null,"approvaldate":null,"isSystemUpload":false,"isDeleted":false,"id":"013475884126683136103","isApproved":null,"orgjoindate":"2022-02-16 09:53:46:511+0000","isSelfDeclaration":false,"updatedBy":null,"orgName":"Tamil Nadu","addedByName":null,"addedBy":null,"associationType":1,"locationIds":["91d9baae-14f1-477a-955c-f91bd9037f0b"],"orgLocation":[{"id":"91d9baae-14f1-477a-955c-f91bd9037f0b","type":"state"}],"externalId":"tn001","userId":"3f8861d6-37a6-493a-8cdc-ca23b27f472c","isSchool":true,"hashTagId":"01269878797503692810","isSSO":true,"isRejected":null,"locations":[{"code":"33","name":"Tamil Nadu","id":"91d9baae-14f1-477a-955c-f91bd9037f0b","type":"state","parentId":null}],"position":null,"orgLeftDate":null},{"organisationId":"01275629929915187240712","approvedBy":null,"channel":"apekx","updatedDate":"2022-02-22 07:41:54:229+0000","approvaldate":null,"isSystemUpload":false,"isDeleted":false,"id":"01275629929915187240712","isApproved":false,"orgjoindate":"2022-02-22 06:22:07:365+0000","isSelfDeclaration":true,"updatedBy":"3f8861d6-37a6-493a-8cdc-ca23b27f472c","orgName":"MPPS NEERPAKOTA Y/C","addedByName":null,"addedBy":"3f8861d6-37a6-493a-8cdc-ca23b27f472c","associationType":2,"locationIds":["bc75cc99-9205-463e-a722-5326857838f8","b5c35cfc-6c1e-4266-94ef-a425c43c7f4e","d8a34179-d925-4bca-a320-3d8c7cbcffa4"],"orgLocation":[{"type":"state","id":"bc75cc99-9205-463e-a722-5326857838f8"},{"type":"district","id":"b5c35cfc-6c1e-4266-94ef-a425c43c7f4e"},{"type":"block","id":"d8a34179-d925-4bca-a320-3d8c7cbcffa4"}],"externalId":"28231601702","userId":"3f8861d6-37a6-493a-8cdc-ca23b27f472c","isSchool":true,"hashTagId":"01275629929915187240712","isSSO":false,"isRejected":false,"locations":[{"code":"28","name":"Andhra Pradesh","id":"bc75cc99-9205-463e-a722-5326857838f8","type":"state","parentId":null},{"code":"2823","name":"CHITTOOR","id":"b5c35cfc-6c1e-4266-94ef-a425c43c7f4e","type":"district","parentId":"bc75cc99-9205-463e-a722-5326857838f8"},{"code":"282316","name":"B.N.KANDRIGA","id":"d8a34179-d925-4bca-a320-3d8c7cbcffa4","type":"block","parentId":"b5c35cfc-6c1e-4266-94ef-a425c43c7f4e"}],"position":null,"orgLeftDate":null}],"provider":null,"countryCode":null,"tncLatestVersionUrl":"https://sunbirdstagingpublic.blob.core.windows.net/termsandcondtions/terms-and-conditions-v12.html","maskedEmail":"ra****@yopmail.com","email":"ra****@yopmail.com","rootOrg":{"keys":{},"organisationSubType":null,"channel":"tn","description":"Tamil Nadu","updatedDate":"2022-02-18 10:07:57:360+0000","organisationType":2,"isTenant":true,"provider":null,"id":"01269878797503692810","email":null,"slug":"tn","isSSOEnabled":null,"orgName":"Tamil Nadu","updatedBy":"fbe926ac-a395-40e4-a65b-9b4f711d7642","locationIds":["91d9baae-14f1-477a-955c-f91bd9037f0b"],"externalId":"tn001","orgLocation":[{"id":"91d9baae-14f1-477a-955c-f91bd9037f0b","type":"state"}],"isRootOrg":true,"rootOrgId":"01269878797503692810","imgUrl":null,"isSchool":true,"homeUrl":null,"createdDate":"2019-02-14 11:41:31:097+0000","createdBy":null,"hashTagId":"01269878797503692810","status":1},"tcUpdatedDate":null,"userLocations":[{"code":"2823","name":"CHITTOOR","id":"b5c35cfc-6c1e-4266-94ef-a425c43c7f4e","type":"district","parentId":"bc75cc99-9205-463e-a722-5326857838f8"},{"code":"28","name":"Andhra Pradesh","id":"bc75cc99-9205-463e-a722-5326857838f8","type":"state","parentId":null},{"code":"282316","name":"B.N.KANDRIGA","id":"d8a34179-d925-4bca-a320-3d8c7cbcffa4","type":"block","parentId":"b5c35cfc-6c1e-4266-94ef-a425c43c7f4e"},{"code":"28231601702","name":"MPPS NEERPAKOTA Y/C","id":"01275629929915187240712","type":"school","parentId":""}],"recoveryPhone":"","userName":"radmin","userId":"3f8861d6-37a6-493a-8cdc-ca23b27f472c","declarations":[],"promptTnC":false,"lastLoginTime":0,"createdDate":"2022-02-16 09:53:45:219+0000","framework":{"board":["State (Tamil Nadu)"],"gradeLevel":["Class 1"],"id":["tn_k-12_5"],"medium":["English"],"subject":["English"]},"createdBy":null,"profileUserType":{"subType":"hm","type":"administrator"},"tncAcceptedVersion":"v12"}}}

**Inside the organisations array in the above response, why both organisation and school data inside organisations array is having "isSchool":true ?**

Please help us understand about this issue, so that we can take next step of actions

Thanks

**Comment by pramodkvarma** [comment]
@vrayulu I am a bit concerned about this. Can you please analyse and let me know your findings? Thanks.

@madhucr

**Reply by vrayulu** [reply]
Sure @pramodkvarma . will check and share details.

**Reply by pramodkvarma** [reply]
Thanks. Looked like an individual user record read API. Why does it return so much details. Maybe needed. Just wanted to make sure and fix any issues as needed.

**Reply by Shakthieshwari** [reply]
@vrayulu Any update on this please ?

Thanks

**Reply by reshmi-nair** [reply]
@Shakthieshwari 
Answer to Q1: User response has the organisation array list to which the user is associated to. School is also considered as an organisation. If you want to select the tenant org from the list, you can consider isSchool=false or there is a separate rootOrg obj itself in the response.

**Reply by pramodkvarma** [reply]
@rhwarrier @vrayulu what's the user consent mechanism for accessing user profile via API?

**Reply by rhwarrier** [reply]
@pramodkvarma  Consent is not factored in for user profile access via API, though we have access control. Using the user read API requires both API token and user token.Only the same user or the org admin of the tenant (who has a token with the admin role) can access the public API to access user details.

**Reply by pramodkvarma** [reply]
That's great! Thanks for the quick response.

Btw, let us look at all APIs where PII data (email etc) may be there and ensure they are fixed if any.

**Reply by aks30** [reply]
> @Shakthieshwari Answer to Q1: User response has the organization array list to which the user is associated to. School is also considered an organization. If you want to select the tenant org from the list, you can consider isSchool=false or there is a separate rootOrg obj itself in the response.

@reshmi-nair - Just for my understanding a user will have only 1 root org, and only max 2 orgs in the organizations list(i.e. 1 root org and 2nd a school if any) not more than that. And any idea why in the 2nd JSON shared for both orgs the isSchool flag is true?

**Reply by vrayulu** [reply]
@aks30 - i think a user can have more than 3 orgs in the list (at least API do not have that kind of restriction, current UI might be limiting to two). 

The 2nd JSON looks like a data issue. @reshmi-nair - please confirm and also share why there is invalid data for this user (it should not be possible via APIs).

**Reply by reshmi-nair** [reply]
@aks30 In 2nd json, the root org/tenant org has organisation type as school. 
@vrayulu Currently the API is allowing to create any organisation type as rootOrg/tenantOrg, as far as the channel provided is unique. There is no restriction specific to organisation type.

**Reply by aks30** [reply]
@Shakthieshwari how do we use the user org value for data reporting? As suggested by @reshmi-nair it is very much possible that a user has a root org which is a school, in that scenario how does it impact us?
